rep-2011.rst

```diff
@@ -0,0 +1,1314 @@
+This service is not expected to be called frequently, and is likely to only occur when new topic or service endpoints are created, and even then, only if the endpoint type hashes do not match.
+
+.. TODO::
+
+    Should each endpoint be asked about their type/version pair or should we assume that the type/version pair guarantees a unique type description and therefore reuse past queries?
+    (wjwwood) I am leaning towards assuming the type name and type version hash combo as being a unique identifier.
+
+Type Description Contents and Format
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The response sent by the ROS Service server will contain a combination of the original ``idl`` or ``msg`` file's content, as well as any necessary information to serialize and deserialize the raw message buffers sent on the topic.
+The response will contain a version of the description that contains comments from the original type description, as those might be relevant to interpreting the semantic meaning of the message fields.
+
+Additionally, the response could include the serialization library used, its version, or any other helpful information from the original producer of the data.
+
+.. TODO::
+
+    What happens if the message consumer doesn't have access to the serialization library stated in the meta-type?
+    (wjwwood) It depends on what you're doing with the response. If you are going to subscribe to a remote endpoint, then I think we just refuse to create the subscription, as communication will not work, and there's a basic assumption that if you're going to communicate with an endpoint then you are using the same or compatible rmw implementations, which includes the serialization technology. The purpose of this section and ROS Service is to provide the information, not to ensure communication can happen.
+
+The ROS 2 message that defines the type description must be able to describe any message type, including itself, and since it is describing the message format, it should work independently from any serialization technologies used.
+This "meta-type description" message would then be used to communicate the structure of the type as part of the "get type description" service response.
+The final form of these interfaces should be found in the reference implementation, but such a Service interface might look like this:
+
+.. code::
+
+    string type_name
+    string type_version_hash
+    ---
+    # True if the type description information is available and populated in the response
+    bool successful
+    # Empty if 'successful' was true, otherwise contains details on why it failed
+    string failure_reason
+
+    # The idl or msg file name
+    string type_description_raw_file_name
+    # The idl or msg file, with comments and whitespace
+    # The file extension and/or the contents can be used to determine the format
+    string type_description_raw
```

Comment:
here is where I see the need to include the raw content of all dependent types as well as the requested type.